DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,350 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+
+// clang-format off
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+// clang-format on
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+using namespace std;
+
+// Forward declare the kernel name in the global scope.
+// This is a FPGA best practice that reduces name mangling in the optimization
+// reports.
+class Add;
+class Div;
+class Mult;
+class ShiftLeft;
+class EfficientShiftLeft;
+class GetBitSlice;
+class SetBitSlice;
+
+using ac_int4 = ac_intN::int4;
+using ac_int14 = ac_intN::int14;
+using ac_int15 = ac_intN::int15;
+using ac_int28 = ac_intN::int28;
+using ac_uint4 = ac_intN::uint4;
+
+void TestAdd(queue &q, const ac_int14 &a, const ac_int14 &b, ac_int15 &c) {
```

Comment:
> This is one of those tutorials that IP authors will look at a lot, so I agree with Abhishek that keeping it simple is best.
Sure, but if I saw this C++ code, I would instantly say that this is way to much boiler plate. The techniques to simplify this are not that complex, I still think we should do it. I can mock up an example of how it would be done.

> Are we using accessors just to be consistent with other code samples? I thought we were switching them all over to USM.
Yes, it is currently still a full stack tutorial. Consistency is more important, we can worry about moving to a new method, across all tutorials, layer.

>I think we should follow Joe's recommendation to use USM and/or device globals for this design so that we get a nice predictable IP interface.
I disagree because of what I said above.